xayn-ai/src/coi/relevance.rs

```diff
@@ -50,18 +50,19 @@ pub(crate) struct Relevances {
 }
 
 impl Relevances {
-    /// Iterates over all tuples in ascending relevance.
-    pub(super) fn iter(
+    /// Iterates over all tuples with matching ids in ascending relevance.
+    pub(super) fn filter(
         &self,
-    ) -> impl Iterator<Item = (CoiId, Relevance, &KeyPhrase)> + DoubleEndedIterator {
-        self.relevance_to_key_phrase
-            .iter()
-            .map(|(&(relevance, coi_id), key_phrases)| {
-                key_phrases
-                    .iter()
-                    .map(move |key_phrase| (coi_id, relevance, key_phrase))
-            })
-            .flatten()
+        coi_ids: impl Clone + Iterator<Item = CoiId>,
+    ) -> impl Iterator<Item = (CoiId, Relevance, &[KeyPhrase])> + DoubleEndedIterator {
```

Comment:
there are situation where `coi_to_relevance` has entries but `relevance_to_key_phrase` is empty, for example if we already called the coi relevance computation but never called the key phrase selection. that's why we still need to filter the `None`s.

on the other hand the `Relevances` are implemented in such a way that if there are relevances in `relevance_to_key_phrase`, then the very same relevances are guaranteed to be present in `coi_to_relevance` as well. so in the end it should turn out the same if we either filter by picking an id from `coi_to_relevance` or if we filter by comparing the ids from `relevance_to_key_phrase`. i guess your suggestion (without the unwrapping) should be faster after all, because we have to do less comparisons, i'll change it accordingly 👌

Reply:
I see, I was hoping that we could avoid that comprising. But thanks for checking